lib/cloud/azure/client_map.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/gravitational/trace"
+)
+
+// ClientMap is a generic map that caches a collection of Azure clients by
+// subscriptions.
+type ClientMap[ClientType any] struct {
+	mu        sync.RWMutex
+	clients   map[string]ClientType
+	newClient func(string, azcore.TokenCredential, *arm.ClientOptions) (ClientType, error)
+}
+
+// NewClientMap creates a new ClientMap.
+func NewClientMap[ClientType any](newClient func(string, azcore.TokenCredential, *arm.ClientOptions) (ClientType, error)) ClientMap[ClientType] {
+	return ClientMap[ClientType]{
+		clients:   make(map[string]ClientType),
+		newClient: newClient,
+	}
+}
+
+// Get returns an Azure client by subscription. A new client is created if the
+// subscription is not found in the map.
+func (m *ClientMap[ClientType]) Get(subscription string, getCredentials func() (azcore.TokenCredential, error)) (client ClientType, err error) {
```

Comment:
Nit: Have you considered moving the `getCredentials` argument to `NewClientMap`? I think the isolation of "initialization" and "usage" would be more apparent. Also, it would look simpler for callers to get clients.

It would look like this:
```go
// Initialize the client map.
cm, err := NewClientMap(newClientFunc, getCredentialsFunc)

// When the caller needs a client, it won't need to provide any initialization dependency.
client, err := cm.Get(subscritpionId)
```

Reply:
yes, I do think what you suggested is cleaner. 

I didn't do it that way because the `NewClientMap` is used in the "constructor" of `clients.NewClients`. At that point, `clients.GetAzureCredential()` cannot be referenced as the object is being created.